### PR TITLE
Improving reproducibility by additional settings in `set_random_seeds`

### PR DESCRIPTION
**Pull request by sliwy**
Last days I faced issues with reproducibility when using `set_random_seeds`. 

First, `set_random_seeds` does not ensure reproducibility because we are not setting `use_deterministic_algorithms()` and we are not setting `torch.backends.cudnn.benchmark = False`. 
Those settings can slow down computations which is a side effect of reproducibility in pytorch (see https://pytorch.org/docs/stable/notes/randomness.html). Still, I feel all people that set random seeds want reproducibility and not just the same weights initialization and batch sampling order.

However some operations may not be possible with deterministic behavior and cuda >= 10.2, so I guess we might want to provide a parameter whether to set it to be deterministic or not (default to deterministic), what do you think? For example it fails in our tests because we use `set_random_seeds` for tests_acceptance, so `use_deterministic_algorithms` affects all tests.

I also added a note about the `PYTHONHASHSEED` setting, which in some cases may be needed to ensure reproducibility. I spent a long time looking for a solution to this problem, so I think it may be good to keep it as a note in the docstring. I had to set `PYTHONHASHSEED` before running scripts which is different than usually suggested `os.environ['PYTHONHASHSEED']`.

This caused me a lot of troubles last week, because I believed that `set_random_seeds` will make my code reproducible :joy:

### CODE REVIEW

**Comment by sliwy** [comment]
I think that we do not see those differences in our CI because we don't use cuda and maybe our datasets are not big enough? (I am not sure regarding the dataset size). I can show you what happens with reproducibility when setting random seeds without `cudnn.benchmark = False`. I am executing on a cluster `plot_sleep_staging.py` example (on my desktop results are exactly the same but I think differences depend on a type of hardware you use). Below training log and confusion matrices (the form is not optimal but I wanted to make it quickly, you should have feeling what's wrong with not setting `cudnn.benchmark`).

## `cudnn.benchmark = False` and same random seed:

 ### First run:
```
epoch    train_bal_acc    train_loss    valid_bal_acc    valid_loss     dur
-------  ---------------  ------------  ---------------  ------------  ------
      1           0.1500        1.6480           0.2236        1.6100  0.3188
      2           0.2021        1.6084           0.2667        1.6095  0.1696
      3           0.2258        1.6001           0.2000        1.6110  0.1670
      4           0.2000        1.5655           0.2000        1.6145  0.1634
      5           0.2100        1.5134           0.2000        1.6331  0.1641
      6           0.2384        1.4419           0.2000        1.7156  0.1633
      7           0.2533        1.3695           0.2000        1.8053  0.1726
      8           0.2835        1.3215           0.2000        1.8395  0.1641
      9           0.3248        1.2944           0.2056        1.8472  0.1730
     10           0.3281        1.2747           0.2203        1.8473  0.1639
[[  0   0   3  58   0]
 [  0   0   8  13   0]
 [  0   0  27 167   0]
 [  0   0   8  67   0]
 [  0   0   4  28   0]]
``` 

### Second run:

```
  epoch    train_bal_acc    train_loss    valid_bal_acc    valid_loss     dur
-------  ---------------  ------------  ---------------  ------------  ------
      1           0.1500        1.6480           0.2236        1.6100  0.3220
      2           0.2021        1.6084           0.2667        1.6095  0.1744
      3           0.2258        1.6001           0.2000        1.6110  0.1726
      4           0.2000        1.5655           0.2000        1.6145  0.1697
      5           0.2100        1.5133           0.2000        1.6331  0.1698
      6           0.2384        1.4420           0.2000        1.7158  0.1700
      7           0.2533        1.3697           0.2000        1.8075  0.1701
      8           0.2835        1.3213           0.2000        1.8381  0.1700
      9           0.3153        1.2950           0.2113        1.8506  0.1724
     10           0.2947        1.2753           0.2553        1.7925  0.1810
[[  0   0   8  53   0]
 [  0   0   9  12   0]
 [  0   0  61 133   0]
 [  0   0  25  50   0]
 [  0   0  12  20   0]]
```

## `cudnn.benchmark = False` and different random seeds:

### First seed:
```
  epoch    train_bal_acc    train_loss    valid_bal_acc    valid_loss     dur
-------  ---------------  ------------  ---------------  ------------  ------
      1           0.1500        1.6480           0.2236        1.6100  0.3003
      2           0.2001        1.6084           0.2667        1.6095  0.2471
      3           0.2278        1.6002           0.2000        1.6112  0.2407
      4           0.2000        1.5653           0.2000        1.6144  0.2394
      5           0.2100        1.5134           0.2000        1.6333  0.2389
      6           0.2384        1.4418           0.2000        1.7156  0.2400
      7           0.2533        1.3697           0.2000        1.8068  0.2398
      8           0.2953        1.3218           0.2000        1.8405  0.2390
      9           0.3204        1.2945           0.2113        1.8492  0.2451
     10           0.3127        1.2740           0.2682        1.7797  0.2415
[[  0   0  12  49   0]
 [  0   0  10  11   0]
 [  0   0  69 125   0]
 [  0   0  30  45   0]
 [  0   0  17  15   0]]
```
### Second seed:
```
  epoch    train_bal_acc    train_loss    valid_bal_acc    valid_loss     dur
-------  ---------------  ------------  ---------------  ------------  ------
      1           0.1515        1.6403           0.2499        1.6105  0.3094
      2           0.2159        1.5812           0.2000        1.6109  0.2465
      3           0.2048        1.5229           0.2037        1.5922  0.2376
      4           0.2092        1.4716           0.2009        1.5920  0.2362
      5           0.2097        1.3990           0.2009        1.5825  0.2369
      6           0.2193        1.3485           0.2018        1.5893  0.2365
      7           0.2373        1.3037           0.2065        1.5896  0.2371
      8           0.2817        1.2681           0.2195        1.5838  0.2369
      9           0.3147        1.2418           0.2520        1.5861  0.2484
     10           0.3304        1.2302           0.3253        1.5654  0.2380
[[ 16   0   4  40   1]
 [  1   0   4  16   0]
 [  4   0  31 157   2]
 [  1   0  14  60   0]
 [  0   0   6  26   0]]
```

## `cudnn.benchmark = False` (Only small difference in the results)
### First run:
```
  epoch    train_bal_acc    train_loss    valid_bal_acc    valid_loss     dur
-------  ---------------  ------------  ---------------  ------------  ------
      1           0.1500        1.6480           0.2236        1.6100  0.2961
      2           0.2001        1.6084           0.2667        1.6095  0.2447
      3           0.2278        1.6002           0.2000        1.6112  0.2403
      4           0.2000        1.5653           0.2000        1.6144  0.2359
      5           0.2100        1.5134           0.2000        1.6333  0.2367
      6           0.2384        1.4418           0.2000        1.7156  0.2387
      7           0.2533        1.3696           0.2000        1.8067  0.2399
      8           0.2953        1.3218           0.2000        1.8406  0.2364
      9           0.3171        1.2945           0.2113        1.8474  0.2395
     10           0.3146        1.2742           0.2664        1.7863  0.2398
[[  0   0  11  50   0]
 [  0   0  10  11   0]
 [  0   0  66 128   0]
 [  0   0  29  46   0]
 [  0   0  17  15   0]]
```

### Second run:
```
  epoch    train_bal_acc    train_loss    valid_bal_acc    valid_loss     dur
-------  ---------------  ------------  ---------------  ------------  ------
      1           0.1500        1.6480           0.2236        1.6100  0.2988
      2           0.2001        1.6084           0.2667        1.6095  0.2498
      3           0.2278        1.6002           0.2000        1.6112  0.2456
      4           0.2000        1.5653           0.2000        1.6144  0.2363
      5           0.2100        1.5135           0.2000        1.6333  0.2391
      6           0.2418        1.4418           0.2000        1.7156  0.2375
      7           0.2533        1.3696           0.2000        1.8061  0.2377
      8           0.2953        1.3219           0.2000        1.8399  0.2372
      9           0.3237        1.2947           0.2113        1.8504  0.2411
     10           0.3011        1.2742           0.2664        1.7833  0.2366
[[  0   0  11  50   0]
 [  0   0  10  11   0]
 [  0   0  67 127   0]
 [  0   0  29  46   0]
 [  0   0  17  15   0]]
```

### Third run:
```
  epoch    train_bal_acc    train_loss    valid_bal_acc    valid_loss     dur
-------  ---------------  ------------  ---------------  ------------  ------
      1           0.1500        1.6480           0.2236        1.6100  0.3003
      2           0.2001        1.6084           0.2667        1.6095  0.2471
      3           0.2278        1.6002           0.2000        1.6112  0.2407
      4           0.2000        1.5653           0.2000        1.6144  0.2394
      5           0.2100        1.5134           0.2000        1.6333  0.2389
      6           0.2384        1.4418           0.2000        1.7156  0.2400
      7           0.2533        1.3697           0.2000        1.8068  0.2398
      8           0.2953        1.3218           0.2000        1.8405  0.2390
      9           0.3204        1.2945           0.2113        1.8492  0.2451
     10           0.3127        1.2740           0.2682        1.7797  0.2415
[[  0   0  12  49   0]
 [  0   0  10  11   0]
 [  0   0  69 125   0]
 [  0   0  30  45   0]
 [  0   0  17  15   0]]
```

# Conclusion:
1. Setting random seed may work well when `cuda` is not used but I haven't checked it yet. On CPU there are some nondeterministic operations that can be selected by torch as well.
2. Setting random seed when using gpu and not setting `cudnn.benchmark =False` may make your work not reproducible at all. Differences are close to running model with different random seeds. On a different problem I observed differences between different random seeds around 15-20% while for the same seed but  `cudnn.benchmark =True` around 10-15% (comparing predictions between them, not the accuracy with true labels).
3. I feel like this function should show a huge warning that if you're using cuda your results won't be reproducible on some devices.
4. Setting random seeds and not setting `cudnn.benchmark` in some cases will make you believe that your computations are reproducible while actually not and this is really bad. Usually people set random seeds for reproducibility and this won't ensure it.

**Comment by robintibor** [comment]
I feel like cudnn.benchmark is something you should know when you are setting it why you set it and that it will result in some nondeterminism. So maybe what we could do is just inside set_random_seeds if cudnn benchmark is set to True, give out a warning that it may not be so well reproducible. Function could also have a flag to suppress that warning. Could be like an argument
`cudnn_benchmark=None`, if it is not set, and it is True, then there is a warning... what do you think?

**Comment by robintibor** [comment]
great maybe add a small note to whats_new?

**Comment by sliwy** [comment]
@robintibor thanks for taking a look at this. 

It's a good idea to have this warning and I think it's enough to somehow warn users about lack of reproducibility that may happen in some cases and give them the solution. 

Sure I'll add a line in whats_new, just waiting for the doc to render, I want to check if everything is ok.

**Comment by sliwy** [comment]
@robintibor I think it's ready, let me know if we need something more here :)

**Comment by sliwy** [comment]
One more thing @robintibor 

Should we remove the line about reproducing results in all the examples? May be misleading if we set `benchmark=True`?

https://github.com/braindecode/braindecode/blob/fa6042b41777b8245cc784da1051d73a974b24a2/examples/plot_sleep_staging.py#L222-L225

**Comment by robintibor** [comment]
We could enhance it to:
```python
# Set random seed to be able to roughly reproduce results 
# Note that with cudnn benchmark set to True, GPU indeterminism
# may still make results substantially differ between runs
 set_random_seeds(seed=random_state, cuda=cuda) 
```

**Comment by sliwy** [comment]
@robintibor done :)

**Comment by robintibor** [comment]
Great that we have a version that works for both of us! Thanks for the work!

File: braindecode/util.py [hunk 1]
```diff
@@ -14,19 +14,46 @@
...
+
     Parameters
     ----------
     seed: int
         Random seed.
     cuda: bool
         Whether to set cuda seed with torch.
+    cudnn_benchmark: bool (default=None)
+        Whether pytorch will use cudnn benchmark. When set to True, results may not be reproducible.
+        When set to False it may slow down computations.
```
Comment (agramfort) [inline braindecode/util.py:31]: can you explain what None will do?

File: examples/plot_bcic_iv_2a_moabb_cropped.py [hunk 2]
```diff
@@ -149,8 +149,10 @@
 device = 'cuda' if cuda else 'cpu'
 if cuda:
     torch.backends.cudnn.benchmark = True
-seed = 20200220  # random seed to make results reproducible
-# Set random seed to be able to reproduce results
+# Set random seed to be able to roughly reproduce results
+# Note that with cudnn benchmark set to True, GPU indeterminism
+# may still make results substantially different between runs
```
Comment (agramfort) [inline examples/plot_bcic_iv_2a_moabb_cropped.py:154]: I would clearly explain that this can be done by setting cudnn_benchmark=True in set_random_seeds

**Review by agramfort — COMMENTED** [review]
sorry to be a bit late to the party

@sliwy do you think it's relevant? if so can you open a new PR to fix this?

🙏
